AK/Utf16View.h

```diff
@@ -0,0 +1,125 @@
+    ~Utf16CodePointIterator() = default;
+
+    bool operator==(const Utf16CodePointIterator& other) const
+    {
+        return (m_ptr == other.m_ptr) && (m_remaining_code_units == other.m_remaining_code_units);
+    }
+
+    bool operator!=(const Utf16CodePointIterator& other) const
+    {
+        return !(*this == other);
+    }
+
+    Utf16CodePointIterator& operator++();
+    u32 operator*() const;
+
+    size_t length_in_code_units() const;
+
+private:
+    Utf16CodePointIterator(const u16* ptr, size_t length)
+        : m_ptr(ptr)
+        , m_remaining_code_units(length)
+    {
+    }
+
+    u16 const* m_ptr { nullptr };
+    size_t m_remaining_code_units { 0 };
+};
+
+class Utf16View {
+public:
+    // Note: transcoding from UTF-8 incurs the cost of storing a Vector of u16 in the view.
+    static Utf16View from_utf8(StringView const&);
+    static Utf16View from_utf8(Utf8View const&);
+
+    static bool is_high_surrogate(u16);
+    static bool is_low_surrogate(u16);
+    static u32 decode_surrogate_pair(u16 high_surrogate, u16 low_surrogate);
+
+    Utf16View() = default;
+    ~Utf16View() = default;
+
+    Utf16View(u16 const* string, size_t length)
+        : m_string(string)
+        , m_length_in_code_units(length)
+    {
+        VERIFY(m_string || (m_length_in_code_units == 0));
+    }
+
+    Utf16View& operator=(Utf16View const&) = delete; // Non-copyable for now, to prevent needlessly copying m_transcoded_data.
+    Utf16View& operator=(Utf16View&&) = default;
+    bool operator==(Utf16View const& other) const;
+
+    enum class AllowInvalidCodeUnits {
+        Yes,
+        No,
+    };
+
+    String to_utf8(AllowInvalidCodeUnits = AllowInvalidCodeUnits::No) const;
+
+    size_t length_in_code_units() const { return m_length_in_code_units; }
+    size_t length_in_code_points() const;
+
+    Utf16CodePointIterator begin() const { return { begin_ptr(), m_length_in_code_units }; }
+    Utf16CodePointIterator end() const { return { end_ptr(), 0 }; }
+
+    u16 const* data() const { return m_string; }
+    u16 code_unit_at(size_t index) const;
+
+    Utf16View substring_view(size_t code_unit_offset, size_t code_unit_length) const;
+    Utf16View substring_view(size_t code_unit_offset) const { return substring_view(code_unit_offset, length_in_code_units() - code_unit_offset); }
+
+    bool validate(size_t& valid_code_units) const;
+    bool validate() const
+    {
+        size_t valid_code_units;
+        return validate(valid_code_units);
+    }
+
+private:
+    explicit Utf16View(Vector<u16> transcoded_data)
+        : m_transcoded_data(move(transcoded_data))
+        , m_string(m_transcoded_data.data())
+        , m_length_in_code_units(m_transcoded_data.size())
+    {
+    }
+
+    u16 const* begin_ptr() const { return m_string; }
+    u16 const* end_ptr() const { return begin_ptr() + m_length_in_code_units; }
+
+    size_t calculate_length_in_code_points() const;
+
+    Vector<u16> m_transcoded_data;
```

Comment:
I don't think making the view own the data is a nice design.
I see two ways forward for this:
- Make this have a `Span<u16 const>+Utf8View+Utf32View` variant (or a bunch of optionals, if that's easier), and transcode things on the fly (this will probably complicate the logic quite a bit for `code_point_at` - you might want to get rid of that and make the user simply iterate over the view, as `Utf8View` does)
- remove this member and require the user to handle the underlying data's lifetime (same as what `Utf32View` does). Your `from_utf8` method up there will have to take a `Vector<u16>&` to dump the code units into.

Either way, let's not make this view own its data.

Reply:
Yeah I didn't particularly like having a "view" actually own any data either...

I was thinking about this overnight, and maybe `JS::Value` should just have a `to_utf16_string` method that returns a `Vector<u16>`. Before starting this, we were under the impression that only a couple of `String.prototype` methods would need UTF-16, but turns out almost all of them do. So then most places `JS::StringPrototype` calls `Value::to_string` would just be flipped to a `to_utf16_string`, and doing 'view' related thing from there is trivial.